walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,162 @@
+    
+    if (available.size() == 0) {
+      return availableWithOptional;
+    }
+    
+    ArrayList<TimeRange> finalTimeRanges = new ArrayList<>();
+
+    for (TimeRange optionalTimeRange: availableWithOptional) {
+      int start = optionalTimeRange.start();
+      int end = optionalTimeRange.end();
+
+      int indexOfIntersection = binarySearchOverlappingTimeRange(available, 0, available.size() - 1, start);
+
+      if (indexOfIntersection != -1) {
+        int mandatoryEnd = available.get(indexOfIntersection).end();
+        if (end <= mandatoryEnd) {
+          finalTimeRanges.add(optionalTimeRange);
+        }
+      }
+    }
+
+    return finalTimeRanges.size() != 0 ? finalTimeRanges : available;
+  }
+    
+  public int binarySearchOverlappingTimeRange(ArrayList<TimeRange> mandatory, int startIndex, int endIndex, int startOfOptional) {
+    int startOfMandatory = mandatory.get(startIndex).start();
+
+    if (startIndex == endIndex) {
+      return startOfMandatory <= startOfOptional ? startIndex : -1;  
+    }
+
+    int midIndex = startIndex + (endIndex - startIndex) / 2;
+
+    if (startOfOptional < mandatory.get(midIndex).start()) {
+      return binarySearchOverlappingTimeRange(mandatory, startIndex, midIndex, startOfOptional);
+    }
+
+    int ret = binarySearchOverlappingTimeRange(mandatory, midIndex + 1, endIndex, startOfOptional);
+
+    return ret == -1 ? midIndex : ret;
+  }
+
+
+  /*
+   * Method to get available times from sorted events
+   * Does this by removing events where there's no attendees, then merging events
+   * which share some time. Finally the gaps are the result.
+   * 
+   * Runtime: O(n)
+   *
+  */
+  public ArrayList<TimeRange> availableTimesFromSortedEvents(HashSet<String> attendees, ArrayList<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    ArrayList<Event> sequentialEvents = new ArrayList<>(events);
+
+    // Removes events in which no attendees intersect between the event and requested attendees
+    for (int i = 0; i < sequentialEvents.size(); ++i) {
+      HashSet<String> attendeesForEvent = new HashSet<String>(sequentialEvents.get(i).getAttendees());
+
+      attendeesForEvent.retainAll(attendees);
+
+      if (attendeesForEvent.size() <= 0) {
+        sequentialEvents.remove(i);
+        --i;
+      }
+    }
+    
+
+    // Merges events which share some time
+    // Ex:       |---A---|
+    //                |--B--|
+    // Becomes:  |----------| <- MERGED_EVENT
+    for (int i = 0; i < sequentialEvents.size() - 1;) {
+      Event eventToCheck = sequentialEvents.get(i);
+      Event nextEvent = sequentialEvents.get(i + 1);
+      
+      TimeRange eventTimeRange = eventToCheck.getWhen();
+      TimeRange nextEventTimeRange = nextEvent.getWhen();
+
+      if (eventTimeRange.overlaps(nextEventTimeRange)) {
+        sequentialEvents.remove(i);
```

Comment:
I'm not sure I can do that. The time range class does not provide any setters so it would be impossible for me to overwrite an element without making changes to the time range class. For the spirit of this exercise I felt that just using the already given functions would be the best decision instead of adding more.

Reply:
Ah yes, it's true that TimeRange is immutable, but sequentialEvents[i] is mutable like `sequentialEvents[i] = some new TimeRange instance` (but actually an iterator now with LinkedList instead of sequentialEvents[i])